### PR TITLE
feat: highlight equal addresses

### PR DESCRIPTION
**Pull request by JoelVR17**
### Feature: Highlight Equal Addresses  |   [Issue #10](https://github.com/walnuthq/op-scan/issues/10)

-------------------------------------------------------------------------------------------------------------------------------------------

*Description*
On a Transaction detail page, when a user hovers over an address, the address should be highlighted in a yellow-ish color. Other addresses on this page that are equal to the address being hovered should also be highlighted.

*Changes*
  1. **Create a AddressProvider and AddressContext.**
    - *File:* AddressContext.tsx
    - *Description:*  Use these functions to provide the status of equal addresses in any part of the project, with `AddressProvider `wrapping the body in layout.tsx.

  2. **Create a AddressLink**
    - *File:* AddressLink.tsx
    - *Description:* 
      - This component is for highlighting equal addresses with a yellow-ish dashed border.
      - If the href prop is an external link, then it shows an icon.

  3. **Implement the AddressLink into latest-transactions**
    - *File:* latest-transactions.tsx
    - *Description:* 
      - Use the AddressLink component in *From/To Addresses* to highlight their equals.

*Notes*
  1. The AddressLink component can be used in any part of the project.

*Images*
  1. HighLight equal addresses *(Dark Theme)*
image

  2. HighLight equal addresses *(Light Theme)*
image

  3. External Icon
image

### CODE REVIEW

File: src/components/pages/home/AddressLink.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+  const handleMouseEnter = () => {
+    dispatch({ type: "SET_HOVERED_ADDRESS", payload: address });
+  };
+
+  const handleMouseLeave = () => {
+    dispatch({ type: "SET_HOVERED_ADDRESS", payload: null });
+  };
+
+  const className = `truncate ${
```
Comment (saimeunt) [inline src/components/pages/home/AddressLink.tsx]: I think the truncate / max-width class names should be moved one level up and we should let the parent component decide how to handle the address styling.
Use an `addressClassName` prop instead and pass it to the underlying span displaying the address.

File: src/components/pages/home/AddressLink.tsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  return (
+    <Link
+      href={href}
+      className={`flex items-center text-primary hover:brightness-150 ${
+        state.hoveredAddress === address ? "text-yellow-200" : ""
+      }`}
+      onMouseEnter={handleMouseEnter}
+      onMouseLeave={handleMouseLeave}
+    >
+      <span className={className}>{children}</span>
```
Comment (saimeunt) [inline src/components/pages/home/AddressLink.tsx]: We need to be able to display the external link icon here, see screenshot attached.

Screenshot 2024-06-24 at 12 28 58

We should display the external link icon only if the href is pointing to an external URL (`new URL(href).origin !== location.origin`)

File: src/components/pages/home/latest-l1-l2-transactions/latest-l1-l2-transaction.tsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+const LatestL1L2Transaction = ({ transaction }: LatestL1L2TransactionProps) => (
+  <div className="flex items-center gap-4 pt-6">
+    <div className="hidden h-9 w-9 items-center justify-center rounded-lg bg-accent text-accent-foreground 2xl:flex">
+      <ReceiptText className="h-5 w-5" />
+    </div>
+    <div className="grid gap-1">
+      <div className="flex items-center gap-1 text-sm">
+        Block#{" "}
+        <span className="flex max-w-60 items-center truncate text-primary md:max-w-96 xl:max-w-60">
```
Comment (saimeunt) [inline src/components/pages/home/latest-l1-l2-transactions/latest-l1-l2-transaction.tsx]: You removed the `a` anchor link here, making the external L1 block unclickable.
Double check you're not introducing regression when submitting PRs.

File: src/components/pages/home/latest-transactions/latest-transaction.tsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+const LatestTransaction = ({ transaction }: LatestTransactionProps) => {
+  return (
+    <div className="flex items-center gap-4 pt-6">
+      <div className="hidden h-9 w-9 items-center justify-center rounded-lg bg-accent text-accent-foreground 2xl:flex">
+        <ReceiptText className="h-5 w-5" />
+      </div>
+      <div className="flex grow flex-col gap-1">
+        <div className="flex items-center gap-1 text-sm">
+          Tx#{" "}
+          <AddressLink
```
Comment (saimeunt) [inline src/components/pages/home/latest-transactions/latest-transaction.tsx]: The hash is not an address and is unique so id does,'t make sense to use the `AddressLink` component.

**Review by saimeunt — CHANGES_REQUESTED** [review]
@JoelVR17 Your AddressLink / AddressContext is looking good, please carefully review the changes requested to merge this PR.
Make sure your code doesn't introduce regressions to the current behavior.
Also, please do not refactor (split lastest-transactions in 2 files) or move things around (latest-l1-tx-batch?) unless this is asked, it only adds confusion to the reviewer.

File: src/components/pages/home/AddressLink.tsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+      ? "max-w-28 md:max-w-48 xl:max-w-28"
+      : type === "latest-transactions" && label === "tx"
+        ? "max-w-60 md:max-w-96 xl:max-w-60"
+        : "max-w-36 md:max-w-48 xl:max-w-36"
+  }`;
+
+  return (
+    <Link
+      href={href}
+      className={`flex items-center text-primary hover:brightness-150 ${
```
Comment (saimeunt) [inline src/components/pages/home/AddressLink.tsx]: Can you try replicating the Etherscan style using tailwind classes?
Something like a yellow dashed border and light yellowish background with high opacity.
Shouldn't be too hard and let's stick to an existing style before coming up with our own.

Screenshot 2024-06-24 at 12 50 10

Comment (JoelVR17) [reply src/components/pages/home/AddressLink.tsx]: Of course, I'll be working on it. Thanks for the feedback!

**Review by saimeunt — APPROVED** [review]

**Pull request merged**